start: Make nCmdShow better match how Windows sets it

squeek502 opened the pull request:
Partial follow up to https://github.com/ziglang/zig/pull/17763#discussion_r1375249495 (cc @e4m2)

`nCmdShow` is a parameter of WinMain that "Controls how the window is to be shown."

From the added comment:

```zig
    // This makes Zig mostly match the nCmdShow behavior of a C program with a WinMain symbol:
    // - With STARTF_USESHOWWINDOW set in STARTUPINFO.dwFlags of the CreateProcess call:
    //   - Compiled with subsystem:console -> nCmdShow is always SW_SHOWDEFAULT
    //   - Compiled with subsystem:windows -> nCmdShow is STARTUPINFO.wShowWindow from
    //     the parent CreateProcess call
    // - With STARTF_USESHOWWINDOW unset:
    //   - nCmdShow is always SW_SHOWDEFAULT
    //
    // TODO: Figure out how to check the subsystem and use it to match the above
    //       behavior, or intentionally avoid matching the above behavior with
    //       regards to the subsystem.
```

With regards to the TODO, it might make sense to just intentionally keep the current behavior: always forward `wShowWindow` if `STARTF_USESHOWWINDOW` is set regardless of the subsystem of the process, since it doesn't seem like a program compiled with the console subsystem would care about `nCmdShow`, or if they do, then I'm not sure why it makes sense to not set `nCmdShow`.

Tested the Windows behavior with the following setup:

---

```c
// launch-show.c
#include <windows.h>
#include <stdio.h>
#include <tchar.h>
#include <shlwapi.h>

void _tmain( int argc, TCHAR *argv[] )
{
    STARTUPINFO si;
    PROCESS_INFORMATION pi;

    ZeroMemory( &si, sizeof(si) );
    si.cb = sizeof(si);
    ZeroMemory( &pi, sizeof(pi) );

    if( argc != 3 )
    {
        printf("Usage: %s [wShowWindow] [cmdline]\n", argv[0]);
        return;
    }

    si.wShowWindow = StrToInt(argv[1]);
    si.dwFlags |= STARTF_USESHOWWINDOW;

    // Start the child process. 
    if( !CreateProcess( NULL,   // No module name (use command line)
        argv[2],        // Command line
        NULL,           // Process handle not inheritable
        NULL,           // Thread handle not inheritable
        FALSE,          // Set handle inheritance to FALSE
        0,              // No creation flags
        NULL,           // Use parent's environment block
        NULL,           // Use parent's starting directory 
        &si,            // Pointer to STARTUPINFO structure
        &pi )           // Pointer to PROCESS_INFORMATION structure
    ) 
    {
        printf( "CreateProcess failed (%d).\n", GetLastError() );
        return;
    }

    // Wait until child process exits.
    WaitForSingleObject( pi.hProcess, INFINITE );

    // Close process and thread handles. 
    CloseHandle( pi.hProcess );
    CloseHandle( pi.hThread );
}
```

Compiled with:

```
zig build-exe launch-show.c -lc -lShlwapi
```

---

```c
// winmain.c
#include <windows.h>
#include <stdio.h>
#include <winbase.h>

int APIENTRY WinMain(_In_ HINSTANCE hInstance,
                     _In_opt_ HINSTANCE hPrevInstance,
                     _In_ LPSTR    lpCmdLine,
                     _In_ int       nCmdShow)
{
    STARTUPINFO info;
    GetStartupInfo(&info);

    CHAR buf[256];
    wsprintf(buf, "WinMain: %d GetStartupInfo: %d", nCmdShow, info.wShowWindow);
    printf("%s\n", buf);
    MessageBoxA(NULL, buf, NULL, 0);
}
```

Compiled with:

```
zig build-exe winmain.c -lc -femit-bin=winmain-console.exe --subsystem console
```
and
```
zig build-exe winmain.c -lc -femit-bin=winmain-windows.exe --subsystem windows
```

---

```
> launch-show.exe 5 winmain-console.exe
WinMain: 10 GetStartupInfo: 5
```
(the STARTUPINFO returned from `GetStartupInfo` has the `wShowWindow` of the CreateProcess call, but `nCmdShow` passed to WinMain is `SW_SHOWDEFAULT`)

```
> launch-show.exe 5 winmain-windows.exe
<message box with the string "WinMain: 5 GetStartupInfo: 5">
```

If the `si.dwFlags |= STARTF_USESHOWWINDOW;` line in `launch-show.exe` is commented out, then `nCmdShow` in `WinMain` is always `SW_SHOWDEFAULT`.

squeek502 commented on lib/std/start.zig:
```diff
@@ -603,14 +603,29 @@ pub fn callMain() u8 {
 
     // second parameter hPrevInstance, MSDN: "This parameter is always NULL"
-    return root.wWinMain(hInstance, null, lpCmdLine, nCmdShow);
+    return root.wWinMain(hInstance, null, lpCmdLine, @intCast(nCmdShow));
```
This `@intCast` is meant to allow for defining `wWinMain` using different types for `nCmdShow`. [The docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-winmain) have `nCmdShow` use the C `int` type, but `ProcessParameters.dwShowWindow` uses the `ULONG` type. This `@intCast` allows for both `c_int` and `u32` to be used when defining a `wWinMain` function on the Zig side.

andrewrk replied:
Instead, you should change the type of `nCmdShow` above to be the minimum integer type in which the value fits. If the user picks an integer type that is smaller than that, it will then correctly be a compile error. For example, you could pick `u31` which will coerce into both `i32` and `u32`, but will be a compile error if the user uses `u8`.

Also, please avoid using `var` unless necessary due to there being a loop.

squeek502 replied:
The [max recognized value is currently `11`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow). And there's multiple different types that it uses in various APIs:

- `u16` in [STARTUPINFO](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfow) (how the 'show window' value is set during CreateProcess)
- `c_int` in [ShowWindows](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow) (what `nCmdShow` is meant to be used with)
- `u32` in `PEB.ProcessParameters.dwShowWindow` (what Windows populates for the process)

so maybe `u16` would make the most sense since `STARTUPINFO` is the bottleneck?

---

Just so I understand better, could you expand a bit on the avoiding `var` thing?

andrewrk commented on lib/std/start.zig:
```diff
@@ -603,14 +603,29 @@ pub fn callMain() u8 {
+    // - With STARTF_USESHOWWINDOW unset:
+    //   - nCmdShow is always SW_SHOWDEFAULT
+    //
+    // TODO: Figure out how to check the subsystem and use it to match the above
+    //       behavior, or intentionally avoid matching the above behavior with
+    //       regards to the subsystem.
```
`std.builtin.subsystem`

squeek502 replied:
Of course! I was way overthinking things.

squeek502 replied:
After trying this, `std.builtin.subsystem` does not actually match up with the subsystem put into the `.exe`, as the linker that writes the OptionalHeader only cares about the `--subsystem` CLI flag, and the `builtin.explicit_subsystem` field that `std.builtin.subsystem` looks for is not set by anything AFAICT.

I've created a new issue for this: https://github.com/ziglang/zig/issues/17808

squeek502 commented:
Addressed the comments, this should be mergeable unless there are more.

I will have a follow-up PR for the subsystem stuff. Right now, `std.builtin.subsystem.? == .Windows` will always be true given how `std.builtin.subsystem` works which is fine (it just means we'll pass on the 'show window' value to processes compiled with the console subsystem).

andrewrk reviewed (approved):
Really appreciate those nice comments, thank you!
